shark-cli/src/main/java/shark/Neo4JCommand.kt

```diff
@@ -0,0 +1,700 @@
+import shark.LeakTraceObject.LeakingStatus.LEAKING
+import shark.LeakTraceObject.LeakingStatus.NOT_LEAKING
+import shark.LeakTraceObject.LeakingStatus.UNKNOWN
+import shark.SharkCliCommand.Companion.echo
+import shark.SharkCliCommand.Companion.retrieveHeapDumpFile
+import shark.SharkCliCommand.Companion.sharkCliParams
+import shark.ValueHolder.BooleanHolder
+import shark.ValueHolder.ByteHolder
+import shark.ValueHolder.CharHolder
+import shark.ValueHolder.DoubleHolder
+import shark.ValueHolder.FloatHolder
+import shark.ValueHolder.IntHolder
+import shark.ValueHolder.LongHolder
+import shark.ValueHolder.ReferenceHolder
+import shark.ValueHolder.ShortHolder
+
+class Neo4JCommand : CliktCommand(
+  name = "neo4j",
+  help = "Convert heap dump to Neo4j database"
+) {
+
+  private val optionalDbFolder by argument("NEO4J_DATABASE_DIRECTORY").file(
+    exists = true,
+    fileOkay = false,
+    folderOkay = true,
+    writable = true
+  ).optional()
+
+  override fun run() {
+    val params = context.sharkCliParams
+    val heapDumpFile = retrieveHeapDumpFile(params)
+
+    val dbFolder = optionalDbFolder ?: heapDumpFile.parentFile
+
+    dump(heapDumpFile, dbFolder, params.obfuscationMappingPath)
+  }
+
+  companion object {
+    val REFERENCE = "REF"
+    val WEAK_REFERENCE = "WEAK_REF"
+    val SOFT_REFERENCE = "SOFT_REF"
+    val PHANTOM_REFERENCE = "PHANTOM_REF"
+
+    fun CliktCommand.dump(
+      heapDumpFile: File,
+      dbParentFolder: File,
+      proguardMappingFile: File?
+    ) {
+      val proguardMapping = proguardMappingFile?.let {
+        ProguardMappingReader(it.inputStream()).readProguardMapping()
+      }
+
+      val name = heapDumpFile.name.substringBeforeLast(".hprof")
+      val dbFolder = File(dbParentFolder, name)
+
+      if (dbFolder.exists()) {
+        val continueImport = TermUi.confirm(
+          "Directory $dbFolder already exists, delete it and continue?",
+          default = true,
+          abort = true
+        ) ?: false
+
+        if (!continueImport) {
+          throw Abort()
+        }
+        echo("Deleting $dbFolder")
+        dbFolder.deleteRecursively()
+      }
+
+      // TODO Support driver + embedded?
+
+      // Get a random, free port. We could do this by using new SocketAddress("localhost", 0)
+      // but we wouldn't be able to retrieve the port via Neo4j means afterwards (it would indicate 0 as ephemeral port)
+
+      // TODO Find an available TCP port instead?
+      val boltListenPort = 2929
+
+      echo("Creating db in $dbFolder")
+      val managementService =
+        DatabaseManagementServiceBuilder(dbFolder.toPath().normalize())
+          .setConfig(BoltConnector.enabled, true)
+          .setConfig(BoltConnector.listen_address, SocketAddress("localhost", boltListenPort))
+          .setConfig(GraphDatabaseSettings.store_internal_log_level, Level.DEBUG)
+          .build()
+      val dbService = managementService.database("neo4j")
+      val api = dbService as GraphDatabaseAPI
```

Comment:
I've been told the only official way to register custom procedures/functions is to use `org.neo4j.configuration.GraphDatabaseSettings#plugin_dir`. Not sure that's gonna work well in this context, but it's worth a try.